README
http://plugins.ardes.com > truncate_html

= truncate_html

Ardes::TruncateHtml is just like the vanilla truncate rails helper, except it respects tags and html entities, and returns valid html.

Rexml and HtmlEntities are used for the parsing.  If the input is invalid html, then hpricot is used to fix it up (but is not used when the input is fine).

Inspired and heavily influenced by http://mikeburnscoder.wordpress.com/2006/11/11/truncating-html-in-ruby/ and the links contained therein.

=== Dependencies

Requires the htmlentities and hpricot gems:

  gem install htmlentities
  gem install hpricot
  
=== Specs and Coverage
* {SPECDOC}[link:files/SPECDOC.html] lists the specifications
* {RSpec Report}[link:rspec_report.html] is a report of spec pass/fails
* {RCov Report}[link:coverage/index.html] is the rcov report

=== Testing
RSpec is used for testing, so the tests are in <tt>spec/</tt> rather than
<tt>test/</tt> Do rake --tasks for more details.

CHANGELOG
* Make API compatible with rails 2.2

* Initial release

spec/helpers/truncate_html_helper_spec.rb
require File.expand_path(File.join(File.dirname(__FILE__), '../spec_helper'))

describe Ardes::TruncateHtmlHelper do
  include Ardes::TruncateHtmlHelper
  
  describe "examples from Rails doc" do
    it "'Once upon a time in a world far far away'" do
      truncate_html("Once upon a time in a world far far away").should == "Once upon a time in a world fa&hellip;"
    end

    it "'Once upon a time in a world far far away', :length => 14" do
      truncate_html("Once upon a time in a world far far away", :length => 14).should == "Once upon a ti&hellip;"
    end

    it "'And they found that many people were sleeping better.', :length => 25, :omission => '(clipped)'" do
      truncate_html("And they found that many people were sleeping better.", :length => 25, :omission => "(clipped)").should == "And they found that many (clipped)"
    end
  end
  
  describe "use cases" do
    def self.with_length_should_equal(n, str)
      it "#{n}, should equal #{str}" do
        truncate_html(@html, :length => n).should == str
      end
    end
  
    describe "html: <p>Hello <strong>World</strong></p>, length: " do
      before { @html = '<p>Hello <strong>World</strong></p>' }
  
      with_length_should_equal 3, '<p>Hel&hellip;</p>'
      with_length_should_equal 7, '<p>Hello <strong>W&hellip;</strong></p>'
      with_length_should_equal 11, '<p>Hello <strong>World</strong></p>'
    end
  
    describe 'html: <p>Hello &amp; <span class="foo">Goodbye</span> <br /> Hi</p>, length: ' do
      before { @html = '<p>Hello &amp; <span class="foo">Goodbye</span> <br /> Hi</p>' }

      with_length_should_equal 7, '<p>Hello &amp;&hellip;</p>'
      with_length_should_equal 9, '<p>Hello &amp; <span class="foo">G&hellip;</span></p>'
      with_length_should_equal 18, '<p>Hello &amp; <span class="foo">Goodbye</span> <br></br> H&hellip;</p>'
    end

    describe '(incorrect) html: <p>Hello <strong>World</p><div>And Hi, length: ' do
      before { @html = '<p>Hello <strong>World</p><div>And Hi' }
  
      with_length_should_equal 10, '<p>Hello <strong>Worl&hellip;</strong></p>'
      with_length_should_equal 30, '<p>Hello <strong>World</strong></p><div>And Hi</div>'
    end
  end
end

garlic.rb
# This is for running specs against target versions of rails
#
# To use do
#   - cp garlic_example.rb garlic.rb
#   - rake get_garlic
#   - [optional] edit this file to point the repos at your local clones of
#     rails, rspec, and rspec-rails
#   - rake garlic:all
#
# All of the work and dependencies will be created in the galric dir, and the
# garlic dir can safely be deleted at any point

garlic do
  repo 'truncate_html', :path => '.'

  repo 'rails', :url => 'git://github.com/rails/rails'
  repo 'rspec', :url => 'git://github.com/dchelimsky/rspec'
  repo 'rspec-rails', :url => 'git://github.com/dchelimsky/rspec-rails'

  ['origin/2-2-stable', 'origin/2-1-stable', 'origin/2-0-stable'].each do |rails|
    target "Rails: #{rails}", :tree_ish => rails do
      prepare do
        plugin 'truncate_html', :clone => true
        plugin 'rspec'
        plugin 'rspec-rails' do
          `script/generate rspec -f`
        end
      end
  
      run do
        cd "vendor/plugins/truncate_html" do
          sh "rake spec:rcov:verify"
        end
      end
    end
  end
end


garlic_example.rb
# This is for running specs against target versions of rails
#
# To use do
#   - cp garlic_example.rb garlic.rb
#   - rake get_garlic
#   - [optional] edit this file to point the repos at your local clones of
#     rails, rspec, and rspec-rails
#   - rake garlic:all
#
# All of the work and dependencies will be created in the galric dir, and the
# garlic dir can safely be deleted at any point

garlic do
  repo 'rails', :url => 'git://github.com/rails/rails'#, :local => "~/dev/vendor/rails"
  repo 'rspec', :url => 'git://github.com/dchelimsky/rspec'#, :local => "~/dev/vendor/rspec"
  repo 'rspec-rails', :url => 'git://github.com/dchelimsky/rspec-rails'#, :local => "~/dev/vendor/rspec-rails"
  repo 'truncate_html', :path => '.'

  target 'edge'
  target '2.0-stable', :branch => 'origin/2-0-stable'
  target '2.0.3', :tag => 'v2.0.3'
  target '2.1-RC1', :tag => 'v2.1.0_RC1'

  all_targets do
    prepare do
      plugin 'truncate_html', :clone => true
      plugin 'rspec'
      plugin('rspec-rails') { sh "script/generate rspec -f" }
    end
  
    run do
      cd("vendor/plugins/truncate_html") { sh "rake spec:rcov:verify" }
    end
  end
end

lib/ardes/truncate_html_helper.rb
require "rexml/parsers/pullparser"
require 'rubygems'
require "htmlentities"
require "hpricot"

module Ardes #:nodoc:
  module TruncateHtmlHelper
    # Truncates html respecting tags and html entities.
    #
    # The API is the same as ActionView::Helpers::TextHelper#truncate.  It uses Rexml for the parsing, and HtmlEntities for entity awareness.  If Rexml raises a ParseException, then Hpricot is used to fixup the tags, and we try again
    #
    # Examples:
    #  truncate_html '<p>Hello <strong>World</strong></p>', :length => 7 # => '<p>Hello <strong>W&hellip;</strong></p>'
    #  truncate_html '<p>Hello &amp; Goodbye</p>', :length => 7          # => '<p>Hello &amp;&hellip;</p>'
    def truncate_html(input, *args)
      # support both 2.2 & earlier APIs
      options = args.extract_options!
      length = options[:length] || args[0] || 30
      omission = options[:omission] || args[1] || '&hellip;'
      
      parser = REXML::Parsers::PullParser.new(input)
      tags, output, chars_remaining = [], '', length
      
      while parser.has_next? && chars_remaining > 0
        element = parser.pull
        case element.event_type
        when :start_element
          output << rexml_element_to_tag(element)
          tags.push element[0]
        when :end_element
          output << "</#{tags.pop}>"
        when :text
          text = HTMLEntities.decode_entities(element[0])
          output << HTMLEntities.encode_entities(text.first(chars_remaining), :named, :basic)
          chars_remaining -= text.length
          output << omission if chars_remaining < 0
        end
      end
      
      tags.reverse.each {|tag| output << "</#{tag}>" }
      output
    
    rescue REXML::ParseException
      truncate_html(Hpricot(input, :fixup_tags => true).to_html, :length => length, :omission => omission)
    end
    
  private
    def rexml_element_to_tag(element)
      "<#{element[0]}#{element[1].inject(""){|m,(k,v)| m << %{ #{k}="#{v}"}} unless element[1].empty?}>"
    end
  end
end